### PR TITLE
Added utility function to find local extrema on a contour map

### LINKED ISSUE #13: Function to find low/high pressure areas in a contour map
Create a utility function that automatically finds local low and high pressure coordinates (denoted by L or H with pressure value as a subscript) on contour plots.

### PR DESCRIPTION
**Pull request by clairefio**
Closes #13

### CODE REVIEW

File: src/util.py [hunk 1]
```diff
@@ -260,12 +265,145 @@ def xr_add_cyclic_longitudes(da, coord):
...
+            xval = x[0]
+            yval = x[1]
+
+            # If the field variable value at the coordinate is less than minVal:
+            if eType == 'Min' and da.data[xval][yval] < minVal:
+                # Add coordinate as an extrema
+                extremacoords.append(tuple(coordarr[xval][yval]))
+            # If the field variable value at the coordinate is greater than maxVal:
+            if eType == 'Max' and da.data[xval][yval] > maxVal:
+                # Add coordinate as an extrema
```
Comment (hCraker) [inline src/util.py:351]: See if you can find a way to produce an error when there aren't any highs or lows that are greater than `maxVal` or less than `minVal`.  The program crashes when that happens and doesn't produce a helpful error message.

File: src/util.py [hunk 2]
```diff
@@ -260,12 +265,145 @@ def xr_add_cyclic_longitudes(da, coord):
 
     return new_da
 
+def findLocalExtrema(da, maxVal=1040, minVal=975, eType='Min'):
```
Comment (hCraker) [inline src/util.py:268]: This is an arbitrary change, but consider changing any use of `max` and `min` in variable names to `high` and `low`. Since our typical user will be in atmospheric or earth sciences, those terms may make more sense given that this function is designed to find highs and lows.

File: src/util.py [hunk 3]
```diff
@@ -260,12 +265,145 @@ def xr_add_cyclic_longitudes(da, coord):
...
+    """
+    Utility function to find local low/high field variable coordinates on a contour map
+
+    Args:
+
+        da: (:class:`xarray.DataArray`):
+            Xarray data array containing the lat, lon, and field variable (ex. pressure) data values
+
+        maxVal (:class:`int`):
```
[the remark this answers is not in the document]
Comment (clairefio) [reply src/util.py:277]: The goal is to have this function be used for mutliple types of data (pressure, temperature, etc.) so I am not going to put a unit.

File: src/util.py [hunk 4]
```diff
@@ -260,12 +265,156 @@ def xr_add_cyclic_longitudes(da, coord):
...
+    point must be greater than highVal, and to classify as a local low, the data point must be less than lowVal.
+
+    Args:
+
+        da: (:class:`xarray.DataArray`):
+            Xarray data array containing the lat, lon, and field variable (ex. pressure) data values
+
+        highVal (:class:`int`):
+            Data value that the local high must be greater than to qualify as a "local high" location.
+            Default highVal is 1040.
```
Comment (hCraker) [inline src/util.py:281]: Don't forget to mention that the defaults are assuming that the data is pressure in hectopascals (hPa)

**Comment by clairefio** [comment]
> Thanks for such a great addition to the library!
> 
> Could you put the list of **geocat-examples** scripts that would use this function? You can put that information in the main description of either this PR or Issue #13 , or both. This will work well for us both to test this function locally and to update those scripts to use this function when this function gets available with a newer version of **geocat-viz** conda package.

The only scripts that would use this function so far are sat_1 and sat_2, but they are still in development. I wanted to wait until this was part of the viz library so I could use this function in those scripts.

**Comment by erogluorhan** [comment]
> The only scripts that would use this function so far are sat_1 and sat_2, but they are still in development. I wanted to wait until this was part of the viz library so I could use this function in those scripts.
 
Since you may have a tight schedule in your remaining time and we need to generate a newer version of the `conda` package of **geocat-viz** (as a new release that probably includes not only this PR but also all the current ongoing PRs of this repo) for you to use this new helper function in **sat_1** and **sat_2**, I can suggest to do the followings: 

1. If the only thing to do with **sat_1** and **sat_2** is to make use of this helper function; instead, define the same helper function in each of those **sat_1** and **sat_2**, make them up and running and ready for a review, and we all can try to merge them if possible
2. Emphasize in the main description of this PR the information that **sat_1** and **sat_2** would use this helper function so that we can keep track of them and update them as soon as we release the newer version of **geocat-viz** via conda, which includes this helper function accessible. 

Please let me know if anything unclear.

File: src/util.py [hunk 5]
```diff
@@ -260,12 +265,156 @@ def xr_add_cyclic_longitudes(da, coord):
...
+
+    """
+    import numpy as np
+    from sklearn.cluster import DBSCAN
+    import warnings
+
+    # Create a 2D array of coordinates in the same shape as the field variable data
+    # so each coordinate is easily mappable to a data value
+    coordarr = []
+    for y in np.array(da.lat):
```
Comment (erogluorhan) [inline src/util.py:306]: Please search for `numpy.meshgrid, and you could solve this task with it because using for-loops are not the most efficient way usually in Python or numpy, especially.

File: src/util.py [hunk 6]
```diff
@@ -260,12 +265,156 @@ def xr_add_cyclic_longitudes(da, coord):
...
+    posfirstzeroes = np.argwhere(arr1 <= bound)
+    negfirstzeroes = np.argwhere(-bound <= arr1)
+
+    # Get all array 2 indexes where gradient value is between -bound and +bound
+    possecondzeroes = np.argwhere(arr2 <= bound)
+    negsecondzeroes = np.argwhere(-bound <= arr2)
+
+    # Find zeroes of all four gradient arrays
+    commonzeroes = []
+    for x in possecondzeroes:
```
Comment (erogluorhan) [inline src/util.py:336]: Similarly to above comment, this for-loop and multiple if-checks inside do not seem to be the most efficient way to do this. Search for `numpy.intersect1d` as it could likely be used in series for solving this problem.

File: src/util.py [hunk 7]
```diff
@@ -260,12 +265,156 @@ def xr_add_cyclic_longitudes(da, coord):
...
+        if x in posfirstzeroes and x in negfirstzeroes and x in negsecondzeroes:
+            commonzeroes.append(x)
+
+    # Find all zeroes that also qualify as low or high values
+    extremacoords = []
+    for x in commonzeroes:
```
Comment (erogluorhan) [inline src/util.py:342]: This `for-loop` and the other `for-loops` throughout the end of the code also look like potentially exchangeable with `numpy`-native methods, too. Please search for `numpy` functionality to handle the task that every single `for-loop` does and try to replace them accordingly.

**Comment by clyne** [comment]
@erogluorhan, I believe at your suggestion that this code is duplicated in https://github.com/NCAR/GeoCAT-examples/pull/108. I've already provided a review of the latter. We should probably consolidate our reviews for the findLocalExtrema function in one place. Would you like that in this PR (I can move my comments) or in https://github.com/NCAR/GeoCAT-examples/pull/108? Thanks

**Comment by erogluorhan** [comment]
> @erogluorhan, I believe at your suggestion that this code is duplicated in [NCAR/GeoCAT-examples#108](https://github.com/NCAR/GeoCAT-examples/pull/108). I've already provided a review of the latter. We should probably consolidate our reviews for the findLocalExtrema function in one place. Would you like that in this PR (I can move my comments) or in [NCAR/GeoCAT-examples#108](https://github.com/NCAR/GeoCAT-examples/pull/108)? Thanks

Thanks for bringing this up @clyne ! @clairefio 's plan was to have this function up and running under **geocat-viz** and then just call it from within [NCAR/GeoCAT-examples#108](https://github.com/NCAR/GeoCAT-examples/pull/108) function. However, the tight schedule made it impossible so that she copied this function's definition to the functions (`NCL_sat_1` and `NCL_sat_2`) in **geocat-examples** in order to at least have the PRs of those functions merged until her departure, if possible. That being said, [NCAR/GeoCAT-examples#108](https://github.com/NCAR/GeoCAT-examples/pull/108) is going to include this function for a temporary purpose until its geocat-viz implementation is finished and a newer version of geocat-viz is released. Since this PR here is aimed at making the utility function as mature as possible before releasing a newer version of geocat-viz, i am slightly in favor of having comments here.

**Comment by clyne** [comment]
@erogluorhan I agree the review for findLocalExtrema should go here. However, it looks like @clairefio already addressed my comments in [GeoCAT-examples](https://github.com/NCAR/GeoCAT-examples/pull/108). How should we proceed to get the functions synced up again?

**Comment by erogluorhan** [comment]
> @erogluorhan I agree the review for findLocalExtrema should go here. However, it looks like @clairefio already addressed my comments in [GeoCAT-examples](https://github.com/NCAR/GeoCAT-examples/pull/108). How should we proceed to get the functions synced up again?

How about the followings:

1. Since we are on the same page, let us review `findLocalExtrema` function-related stuff under this PR rather than geocat-examples from now on.

2. Let us ask to @clairefio (with my apologies) to sync her latest version of `findLocalExtrema` from geocat-examples PR # [108](https://github.com/NCAR/GeoCAT-examples/pull/108) into this PR here, and we review those changes here rather than under geocat-examples?

3. Let us keep reviewing  `findLocalExtrema` only here and `NCL_sat_1` in **geocat-examples** until @clairefio 's departure. In case these works are not finished by the last day, she may simply copy the latest version of `findLocalExtrema` from here to under her relevant **geocat-examples** PRs for sync'ing purposes.

4. Let us have @michaelavs in the loop as she would likely be taking over these works in case they are not finished until @clairefio  leaves.

Please anybody in the loop let us know if this does not make sense.

**Comment by clyne** [comment]
> > @erogluorhan I agree the review for findLocalExtrema should go here. However, it looks like @clairefio already addressed my comments in [GeoCAT-examples](https://github.com/NCAR/GeoCAT-examples/pull/108). How should we proceed to get the functions synced up again?
> 
> How about the followings:
> 
> 1. Since we are on the same page, let us review `findLocalExtrema` function-related stuff under this PR rather than geocat-examples from now on.
> 2. Let us ask to @clairefio (with my apologies) to sync her latest version of `findLocalExtrema` from geocat-examples PR # [108](https://github.com/NCAR/GeoCAT-examples/pull/108) into this PR here, and we review those changes here rather than under geocat-examples?
> 3. Let us keep reviewing  `findLocalExtrema` only here and `NCL_sat_1` in **geocat-examples** until @clairefio 's departure. In case these works are not finished by the last day, she may simply copy the latest version of `findLocalExtrema` from here to under her relevant **geocat-examples** PRs for sync'ing purposes.
> 4. Let us have @michaelavs in the loop as she would likely be taking over these works in case they are not finished until @clairefio  leaves.
> 
> Please anybody in the loop let us know if this does not make sense.

That all sounds good, with apologies to @clairefio :-)

**Comment by clairefio** [comment]
@clyne @erogluorhan I took out a lot of the for-loops, as well as the gradient function/bounds checking (which I found I actually didn't need). I also added in default arguments that set the high/low threshold values to "extreme" values. The default high pressure threshold is 0, which means that any pressure value above 0 is qualified to be a local high pressure. The default low pressure threshold is 1000, so anything below that qualifies to be a local low pressure. So, the user doesn't have to set a high/low value and they will still get extrema that make sense. This is an example run with the user not specifying high/low values in the function call.

default args

File: src/util.py [hunk 8]
```diff
@@ -260,12 +265,110 @@ def xr_add_cyclic_longitudes(da, coord):
...
+            Determines which extrema are being found- minimum or maximum, respectively.
+            Default eType is 'Low'.
+    Returns:
+        clusterExtremas (:class:`list`):
+            List of coordinate tuples in GPS form (lon in degrees, lat in degrees)
+            that specify local low/high locations
+    """
+    
+    import numpy as np
+    from sklearn.cluster import DBSCAN
```
Comment (erogluorhan) [inline src/util.py:294]: You should add `scikit-learn` dependency into **setup.py** in the root directory.

Comment (clairefio) [reply src/util.py:294]: Done!

Comment (erogluorhan) [reply src/util.py:294]: Looks good! As soon as @clyne approves, we will merge this PR.
